misc/docs-linter/links.go

```diff
@@ -3,105 +3,73 @@ package main
 import (
 	"bufio"
 	"bytes"
-	"context"
 	"fmt"
-	"golang.org/x/sync/errgroup"
-	"io"
-	"mvdan.cc/xurls/v2"
-	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
 	"strings"
-	"sync"
 )
 
-// extractUrls extracts URLs from a file and maps them to the file
-func extractUrls(fileContent []byte) []string {
+// extractLocalLinks extracts links to local files given file content
+func extractLocalLinks(fileContent []byte) []string {
 	scanner := bufio.NewScanner(bytes.NewReader(fileContent))
-	urls := make([]string, 0)
+	links := make([]string, 0)
+	// Regular expression to match markdown links
+	re := regexp.MustCompile(`]\((\.\.?/.+?)\)`)
 
 	// Scan file line by line
 	for scanner.Scan() {
 		line := scanner.Text()
 
-		// Extract links
-		rxStrict := xurls.Strict()
-		url := rxStrict.FindString(line)
+		// Find all matches
+		matches := re.FindAllString(line, -1)
 
-		// Check for empty links and skip them
-		if url == " " || len(url) == 0 {
-			continue
-		}
+		// Extract and print the local file links
+		for _, match := range matches {
+			// Remove ]( from the beginning and ) from end of link
+			match = match[2 : len(match)-1]
+
+			// Remove markdown headers in links
+			pos := strings.Index(match, "#")
+
+			if pos != -1 {
+				match = match[:pos]
+			}
+
+			pos = strings.Index(match, " ")
```

Comment:
In what situations is there a space in the link?

Reply:
For example, the way we do [embedmd links](https://github.com/gnolang/gno/blob/bedd0f9171d9a3ef66a0afb955842abdfe07f21e/docs/concepts/packages.md?plain=1#L23). Kind of an edge case, but it needed to be parsed out, hence this code.

Reply:
Okay thanks.  Can you put a comment there explaining why it is necessary? I'm not sure most readers would understand why this is necessary.

Reply:
I've played around with this a bit an realized that there was a bug in the code; I've separated the embedmd link check out, and added proper tests for it: [2061e51](https://github.com/gnolang/gno/pull/2416/commits/2061e516d89464d643852f58d337ad8195cd6c67)